scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/action/CommonModelDataInfoProvider.scala

```diff
@@ -26,18 +31,65 @@ abstract class CommonModelDataInfoProvider(modelData: ModelData) {
     modelData.modelClassLoader,
     Seq.empty,
     ProductionServiceInvocationCollector,
-    ComponentUseCase.TestDataGeneration,
+    RuntimeMode.Live,
+    NodesDeploymentData.empty,
     nonServicesLazyParamStrategy = LazyParameterCreationStrategy.default
   )
 
-  protected def compileSourceNode(
+  private lazy val subGraphCompiler = new PartSubGraphCompiler(nodeCompiler)
+
+  private lazy val scenarioCompiler = new ProcessCompiler(
+    modelData.modelClassLoader,
+    subGraphCompiler,
+    GlobalVariablesPreparer(modelData.modelDefinitionWithClasses.modelDefinition.expressionConfig),
+    nodeCompiler,
+    modelData.customProcessValidator
+  )
+
+  def collectAllSources(scenario: CanonicalProcess): List[SourceNodeData] = {
+    scenario.collectAllNodes.flatMap(asSource) ++ scenario.collectAllNodes.flatMap(asFragmentInputDefinition)
+  }
+
+  def compileSourceNode(
       source: SourceNodeData
   )(implicit jobData: JobData, nodeId: NodeId): ValidatedNel[ProcessCompilationError, Source] = {
     nodeCompiler.compileSource(source).compiledObject
   }
 
-  protected def collectAllSources(scenario: CanonicalProcess): List[SourceNodeData] = {
-    scenario.collectAllNodes.flatMap(asSource) ++ scenario.collectAllNodes.flatMap(asFragmentInputDefinition)
+  def compileScenario(scenario: CanonicalProcess)(implicit jobData: JobData): Map[NodeComponentInfo, Any] = {
+    scenarioCompiler.compile(scenario).result match {
+      case Validated.Valid(compiledScenario: CompiledProcessParts) =>
+        extractComponents(compiledScenario.sources.toList)
+      case _ => Map.empty
+    }
   }
 
+  private def extractComponents(parts: List[ProcessPart])(implicit jobData: JobData): Map[NodeComponentInfo, Any] =
+    parts.foldLeft(Map.empty[NodeComponentInfo, Any]) { (acc, part) =>
+      val nodeComponentInfo = NodeComponentInfoExtractor.fromScenarioNode(part.node.data)
+      part match {
+        case source: SourcePart =>
+          acc + (nodeComponentInfo -> source.obj) ++ compilePart(source) ++ extractComponents(source.nextParts)
+        case custom: CustomNodePart =>
+          acc + (nodeComponentInfo -> custom.transformer) ++ compilePart(custom) ++ extractComponents(custom.nextParts)
+        case sink: SinkPart =>
+          acc + (nodeComponentInfo -> sink.obj)
+      }
+    }
+
+  private def compilePart(part: ProcessPart)(implicit jobData: JobData): Map[NodeComponentInfo, Any] =
```

Comment:
The same as above, we need to emphasize that this compilation is for the purpose of collection of all custom nodes. Maybe the same name `compileAllCustomNodes`? Or `compileAllPartCustomNodes`

Reply:
Done